### PR TITLE
[MI-2523] Refactored the unit tests

### PR DESCRIPTION
**Pull request by manojmalik20**
Modified some constants and created new ones
Modified all the APIs to improve them
Created several new utils for testing APIs with middlewares
Added new tests for the middlewares
Removed unnecessary tests

### CODE REVIEW

File: server/plugin/api_test.go [hunk 1]
```diff
@@ -36,17 +37,39 @@ func (ph panicHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
 
+func setupTestPlugin(api *plugintest.API, store *mock_plugin.MockStore) (*Plugin, *plugintest.API) {
+	p := &Plugin{}
+	path, _ := filepath.Abs("../..")
+	api.On("GetBundlePath").Return(path, nil)
+	p.SetAPI(api)
+	if store != nil {
+		p.store = store
+	}
+	p.router = p.initializeAPI()
```
Comment (ayusht2810) [inline server/plugin/api_test.go]:
```suggestion
	}
	
	p.router = p.initializeAPI()
```

File: server/plugin/api_test.go [hunk 2]
```diff
@@ -36,17 +37,39 @@ func (ph panicHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+	p.setConfiguration(
+		&configuration{
+			ServiceNowURL:               "mockURL",
+			ServiceNowOAuthClientID:     "mockCLientID",
```
Comment (ayusht2810) [inline server/plugin/api_test.go]:
```suggestion
			ServiceNowOAuthClientID:     "mockClientID",
```

File: server/plugin/api_test.go [hunk 3]
```diff
@@ -823,9 +679,9 @@ func TestPlugin_handlePickerSelection(t *testing.T) {
 			expectedResponse: testutils.ExpectedResponse{
 				StatusCode: http.StatusOK,
 			},
-			callError:         errors.New("mockError"),
-			ParseAuthTokenErr: nil,
-			LoadUserErr:       nil,
+			setupClient: func(c *mock_plugin.MockClient) {
+				c.EXPECT().SendMessageToVirtualAgentAPI(testutils.GetServiceNowSysID(), "mockOption", true, &serializer.MessageAttachment{}).Return(errors.New("mockError"))
```
Comment (ayusht2810) [inline server/plugin/api_test.go]: we can add some meaningful error here

File: server/plugin/api_test.go [hunk 4]
```diff
@@ -1418,3 +1163,91 @@ func TestPlugin_handleFileAttachments(t *testing.T) {
...
+
+		result := w.Result()
+		require.NotNil(t, result)
+		defer result.Body.Close()
+
+		assert.Equal(http.StatusUnauthorized, result.StatusCode)
+		var resp *serializer.APIErrorResponse
+		err := json.NewDecoder(result.Body).Decode(&resp)
+		require.Nil(t, err)
+
```
Comment (ayusht2810) [inline server/plugin/api_test.go]:
```suggestion
```

File: server/plugin/api_test.go [hunk 5]
```diff
@@ -1418,3 +1163,91 @@ func TestPlugin_handleFileAttachments(t *testing.T) {
...
+func TestCheckOAuth(t *testing.T) {
+	requestURL := fmt.Sprintf("%s%s", pathPrefix, constants.PathSetDateTimeDialog)
+	for name, test := range map[string]struct {
+		SetupAPI             func(*plugintest.API)
+		SetupPluginAndStore  func(p *Plugin, s *mock_plugin.MockStore)
+		ExpectedStatusCode   int
+		ExpectedErrorMessage string
+	}{
+		"failed to load user": {
```
Comment (ayusht2810) [inline server/plugin/api_test.go]:
```suggestion
		"failed to load the user": {
```

File: server/plugin/api_test.go [hunk 6]
```diff
@@ -1418,3 +1163,91 @@ func TestPlugin_handleFileAttachments(t *testing.T) {
...
+			result := w.Result()
+			require.NotNil(t, result)
+			defer result.Body.Close()
+
+			assert.Equal(test.ExpectedStatusCode, result.StatusCode)
+			if test.ExpectedErrorMessage != "" {
+				var resp *serializer.APIErrorResponse
+				err := json.NewDecoder(result.Body).Decode(&resp)
+				require.Nil(t, err)
+
```
Comment (ayusht2810) [inline server/plugin/api_test.go]:
```suggestion
```

File: server/plugin/oauth_test.go [hunk 7]
```diff
@@ -132,19 +94,7 @@ func TestPlugin_httpOAuth2Complete(t *testing.T) {
...
-			httpTest: httpTestJSON,
-			request: testutils.Request{
-				Method: http.MethodGet,
-				URL:    "/api/v1/oauth2/complete?code=mockCode&state=mockState",
-			},
-			expectedResponse: testutils.ExpectedResponse{
-				StatusCode: http.StatusUnauthorized,
-			},
-			userID: "",
+			completeOAuthErr: nil,
```
Comment (ayusht2810) [inline server/plugin/oauth_test.go]:
```suggestion
```

Comment (ayusht2810) [reply server/plugin/oauth_test.go]: check everywhere

File: server/plugin/oauth_test.go [hunk 8]
```diff
@@ -40,19 +39,6 @@ func TestPlugin_httpOAuth2Connect(t *testing.T) {
...
-			httpTest: httpTestJSON,
-			request: testutils.Request{
-				Method: http.MethodGet,
-				URL:    "/api/v1/oauth2/connect",
-			},
-			expectedResponse: testutils.ExpectedResponse{
-				StatusCode: http.StatusUnauthorized,
-			},
-			userID:        "",
 			InitOAuth2Err: nil,
```
Comment (ayusht2810) [inline server/plugin/oauth_test.go]: same

File: server/plugin/api_test.go [hunk 9]
```diff
@@ -36,17 +37,39 @@ func (ph panicHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+			ServiceNowOAuthClientID:     "mockCLientID",
+			ServiceNowOAuthClientSecret: "mockClientSecret",
+			EncryptionSecret:            "mockEncryptionSecret",
+			WebhookSecret:               "mockWebhookSecret",
+			MattermostSiteURL:           "mockSiteURL",
+			PluginID:                    "mockPluginID",
+			PluginURL:                   "mockPluginURL",
+			PluginURLPath:               "mockPluginURLPath",
+		})
+	return p, api
```
Comment (ayusht2810) [inline server/plugin/api_test.go]:
```suggestion
		})
		
	return p, api
```

**Review by ayusht2810 — APPROVED** [review]

**Review by Kshitij-Katiyar — APPROVED** [review]

**Pull request merged**